clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslAuthenticatorTest.java

```diff
@@ -714,6 +753,265 @@ public void testJaasConfigurationForListener() throws Exception {
+        verifyFailureInteropWithoutSaslAuthenticateHeader(false, true, SecurityProtocol.SASL_PLAINTEXT, "SCRAM-SHA-256");
+    }
+
+    /**
+     * Tests SASL/SCRAM authentication failure over PLAINTEXT with old version of client
+     * that does not support SASL_AUTHENTICATE headers and new version of server.
+     */
+    @Test
+    public void oldSaslScramPlaintextClientWithoutSaslAuthenticateHeaderFailure() throws Exception {
+        verifyFailureInteropWithoutSaslAuthenticateHeader(true, false, SecurityProtocol.SASL_PLAINTEXT, "SCRAM-SHA-256");
+    }
+
+    /**
+     * Tests SASL/PLAIN authentication failure over SSL with old version of server
+     * that does not support SASL_AUTHENTICATE headers and new version of client.
+     */
+    @Test
+    public void oldSaslPlainSslServerWithoutSaslAuthenticateHeaderFailure() throws Exception {
+        verifyFailureInteropWithoutSaslAuthenticateHeader(false, true, SecurityProtocol.SASL_SSL, "PLAIN");
+    }
+
+    /**
+     * Tests SASL/PLAIN authentication failure over SSL with old version of client
+     * that does not support SASL_AUTHENTICATE headers and new version of server.
+     */
+    @Test
+    public void oldSaslPlainSslClientWithoutSaslAuthenticateHeaderFailure() throws Exception {
+        verifyFailureInteropWithoutSaslAuthenticateHeader(true, false, SecurityProtocol.SASL_SSL, "PLAIN");
+    }
+
+    /**
+     * Tests SASL/SCRAM authentication failure over SSL with old version of server
+     * that does not support SASL_AUTHENTICATE headers and new version of client.
+     */
+    @Test
+    public void oldSaslScramSslServerWithoutSaslAuthenticateHeaderFailure() throws Exception {
+        verifyFailureInteropWithoutSaslAuthenticateHeader(false, true, SecurityProtocol.SASL_SSL, "SCRAM-SHA-512");
+    }
+
+    /**
+     * Tests SASL/SCRAM authentication failure over SSL with old version of client
+     * that does not support SASL_AUTHENTICATE headers and new version of server.
+     */
+    @Test
+    public void oldSaslScramSslClientWithoutSaslAuthenticateHeaderFailure() throws Exception {
+        verifyFailureInteropWithoutSaslAuthenticateHeader(true, false, SecurityProtocol.SASL_SSL, "SCRAM-SHA-512");
+    }
+
+    private void verifyInteropWithoutSaslAuthenticateHeader(boolean enableHeaderOnServer, boolean enableHeaderOnClient,
+            SecurityProtocol securityProtocol, String saslMechanism) throws Exception {
+        configureMechanisms(saslMechanism, Arrays.asList(saslMechanism));
+        createServer(securityProtocol, saslMechanism, enableHeaderOnServer);
+
+        String node = "0";
+        createClientConnection(securityProtocol, saslMechanism, node, enableHeaderOnClient);
+        NetworkTestUtils.checkClientConnection(selector, "0", 100, 10);
+    }
+
+    private void verifyFailureInteropWithoutSaslAuthenticateHeader(boolean enableHeaderOnServer, boolean enableHeaderOnClient,
```

Comment:
Ditto as the above. It seems the method supports with or w/o SaslAuthenticateHeader.